src/GitVersion.Core/Core/RepositoryStore.cs

```diff
@@ -239,52 +239,60 @@ public IEnumerable<IBranch> ExcludingBranches(IEnumerable<IBranch> branchesToExc
         // TODO Should we cache this?
         public IEnumerable<IBranch> GetBranchesContainingCommit(ICommit commit, IEnumerable<IBranch> branches = null, bool onlyTrackedBranches = false)
         {
-            branches ??= repository.Branches.ToList();
-            static bool IncludeTrackedBranches(IBranch branch, bool includeOnlyTracked) => includeOnlyTracked && branch.IsTracking || !includeOnlyTracked;
-
             if (commit == null)
             {
                 throw new ArgumentNullException(nameof(commit));
             }
 
-            using (log.IndentLog($"Getting branches containing the commit '{commit.Id}'."))
+            return InnerGetBranchesContainingCommit();
+
+            static bool IncludeTrackedBranches(IBranch branch, bool includeOnlyTracked) => includeOnlyTracked && branch.IsTracking || !includeOnlyTracked;
+
+            // Yielding part is split from the main part of the method to avoid having the exception check performed lazily.
+            // Details at https://github.com/GitTools/GitVersion/issues/2755
+            IEnumerable<IBranch> InnerGetBranchesContainingCommit()
```

Comment:
I could also make this local function static and pass-in required arguments. The performance diff is minimal enough to be more of a stylistic choice than anything else.

Reply:
Yes, that would be less exotic and more approachable, I think. 👍🏼